pmlc/conversion/tile_to_pxa/tile_to_pxa.cc

```diff
@@ -1362,35 +1364,107 @@ struct ScatterOpConversion : public OpConversionPattern<tile::ScatterOp> {
-        rewriter.create<pxa::PxaLoadOp>(loc, indices, idxLoadMap, idxLoadOps)
-            .getResult();
+      Value indexVal =
+          rewriter.create<pxa::PxaLoadOp>(loc, indices, idxLoadMap, idxLoadOps)
+              .getResult();
 
-    // Cast the index value from its integer type to the index type
-    if (!indexVal.getType().isa<IndexType>()) {
-      // cast from whatever integer type it has to index type
-      auto indexType = rewriter.getIndexType();
-      indexVal = rewriter.create<mlir::IndexCastOp>(loc, indexVal, indexType)
-                     .getResult();
-    }
+      // Cast the index value from its integer type to the index type
+      if (!indexVal.getType().isa<IndexType>()) {
+        // cast from whatever integer type it has to index type
+        auto indexType = rewriter.getIndexType();
+        indexVal = rewriter.create<mlir::IndexCastOp>(loc, indexVal, indexType)
+                       .getResult();
+      }
 
-    // Combine the index value with the loop dimension indexes to create the
-    // destination affine map.
-    size_t dstDims = resultMemRefType.getShape().size();
-    SmallVector<Value, 4> dstOps;
-    dstOps.push_back(indexVal);
-    for (size_t i = 1; i < dstDims; ++i) {
-      dstOps.push_back(loop.getIVs()[i]);
+      // Combine the index value with the loop dimension indexes to create the
+      // destination affine map.
+      for (int i = 0; i < axis; ++i) {
+        dstOps.push_back(loop.getIVs()[i]);
+      }
+
+      for (size_t i = idxStart; i < srcDims; ++i) {
+        dstOps.push_back(loop.getIVs()[i]);
+      }
+
+      dstOps[axis] = indexVal;
+    } else {
+      std::vector<Value> idxs, combIdx(idxDims);
+      for (size_t i = 0; i < idxDims - 1; ++i) {
+        combIdx[i] = loop.getIVs()[i];
+      }
+      for (int64_t i = 0; i < idxShape[idxDims - 1]; ++i) {
+        combIdx[idxDims - 1] = rewriter.create<mlir::ConstantIndexOp>(loc, i);
+        Value indexVal =
+            rewriter.create<pxa::PxaLoadOp>(loc, indices, combIdx).getResult();
+
+        // Cast the index value from its integer type to the index type
+        if (!indexVal.getType().isa<IndexType>()) {
+          // cast from whatever integer type it has to index type
+          auto indexType = rewriter.getIndexType();
+          indexVal =
+              rewriter.create<mlir::IndexCastOp>(loc, indexVal, indexType)
+                  .getResult();
+        }
+        idxs.push_back(indexVal);
+      }
+      dstOps.insert(dstOps.begin(), idxs.begin(), idxs.end());
+      for (size_t i = idxDims - 1; i < srcDims; ++i) {
+        dstOps.push_back(loop.getIVs()[i]);
+      }
     }
 
+    if (op.updateMode() != ScatterUpdateMode::none) {
+      auto elementType = data.getType().cast<MemRefType>().getElementType();
+      auto zeroVal = rewriter.create<mlir::ConstantOp>(
+          loc, elementType, rewriter.getFloatAttr(elementType, 0.0));
+
+      rewriter.create<mlir::StoreOp>(loc, zeroVal, data, dstOps);
```

Comment:
This changes the input date... We should keep the input data unchanged, shouldn't we?

Reply:
Another thing is should Update mode always assume indices has no duplicates?

Reply:
Yes. Maybe just copying the input to the output is enough.

Reply:
> Another thing is should Update mode always assume indices has no duplicates? Maybe for update mode we should just override for duplicate indices instead of sum.

Yes but the problem is that updating can be happening on multiple threads (if this is in a parallel op or something) so which one will be overwrriten is **UNDEFINED**.